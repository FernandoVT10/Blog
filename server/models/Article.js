import mongoose, { Schema } from "mongoose";

const article = new Schema({
    title: {
        type: String,
        maxlength: 50,
        required: true
    },
    description: {
        type: String,
        maxlength: 250,
        required: true
    },
    cover: {
        type: String,
        required: true
    },
    content: {
        type: String,
        required: true
    },
    categories: [
        {
            type: mongoose.Types.ObjectId,
            ref: "categories"
        }
    ],
    dayViews: {
        type: Number,
        default: 0
    },
    monthViews: {
        type: Number,
        default: 0
    },
    totalViews: {
        type: Number,
        default: 0
    }
}, { timestamps: true });

export default mongoose.model("articles", article);