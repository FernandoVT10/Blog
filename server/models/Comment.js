import mongoose, { Schema } from "mongoose";

const comment = new Schema({
    articleId: {
        type: mongoose.Types.ObjectId,
        ref: "articles"
    },
    name: {
        type: String,
        maxlength: 30,
        required: true
    },
    comment: {
        type: String,
        maxlength: 500,
        required: true
    }
}, { timestamps: true });

export default mongoose.model("comments", comment)