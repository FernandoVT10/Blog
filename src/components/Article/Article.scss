@use "../../styles/variables";

.article {
    background: variables.$primary;
    padding: 0;
    color: white;

    &__cover {
        width: 100%;
        height: 400px;
        background-position: center !important;
        background-attachment: fixed !important;
        background-size: cover !important;

        &--edit {
            position: relative;
            display: flex;
            justify-content: center;
            align-items: center;

            &::before {
                content: "";
                opacity: 0;
                position: absolute;
                width: 100%;
                height: 100%;
                background: rgba($color: #000000, $alpha: .8);
                transition: .3s;
            }
        }

        &__label {
            cursor: pointer;
            position: relative;
            opacity: 0;
            z-index: 2;
            color: white;
            font-size: 30px;
            padding: 15px 40px;
            background: variables.$primary;
            transition: .3s;
        }

        &__input {
            display: none;
        }

        &:hover {
            &::before {
                opacity: 1;
            }

            & .article__cover__label {
                opacity: 1;

                &:hover {
                    opacity: .6;
                }
            }
        }
    }

    &__title {
        font-weight: 600;
        font-size: 30px;
    }

    &__edit-title {
        outline: 0;
        width: 100%;
        font-weight: 600;
        font-size: 30px;
        border: none;
        color: white;
        height: 50px;
        resize: none;
        word-break: break-all;
        background: transparent;

        &-error {
            display: none;
            color: variables.$danger;
            margin-bottom: 15px;
        }

        &:invalid {
            box-shadow: 0 0px 2px 1px variables.$danger;

            & ~ .article__edit-title-error {
                display: block;
            }
        }
    }

    &__content-container {
        padding: 30px;
    }

    &__content {
        margin-top: 30px;
    }

    &__content-editor {
        margin-bottom: 15px;

        &.ql-disabled {
            border: none !important;
        }
    }

    &__description {
        width: 100%;
        padding: 20px;
        border-radius: 5px;
        background: transparent;
        color: white;
    }

    &__categories {
        margin-top: 10px;
        font-weight: bold;
        font-size: 14px;
        color: variables.$secondary;

        &__container {
            display: inline;
        }

        &__button-container {
            display: inline-block;
            position: relative;
            margin: 0 5px;
        }

        &__add-category {
            z-index: 10;
            position: absolute;
            display: none;
            top: 30px;
            left: 0;
            flex-direction: column;
            background: variables.$black;
            border: solid 2px variables.$secondary;
            padding: 10px 20px;
            max-height: 160px;
            width: 180px;
            overflow: auto;

            &--active {
                display: flex;
            }

            &::-webkit-scrollbar {
                width: 3px;
                background: variables.$black;
            }

            &::-webkit-scrollbar-thumb {
                background: variables.$secondary;
            }
        }

        &__add-button {
            position: relative;
            outline: 0 !important;
            background: transparent;
            font-size: 12px;
            border: none;
            border-radius: 5px;
            color: variables.$secondary;
            transition: .3s;

            &:hover {
                color: white;
            }
        }
    }

    &__category {
        display: inline-block;
        color: variables.$secondary;
        transition: .3s;

        &:hover {
            text-decoration: none;
            color: white;
        }
    }

    &__buttons-container {
        position: fixed;
        z-index: 100;
        right: 25px;
        bottom: 25px;
    }

    &__float-button {
        outline: 0 !important;
        display: inline-flex;
        margin: 10px;
        justify-content: center;
        align-items: center;
        background: variables.$secondary;
        width: 50px;
        height: 50px;
        border-radius: 50%;
        border: none;
        transition: .3s;

        &--cancel {
            background: variables.$danger;
            color: white;
        }

        &:hover {
            opacity: .8;
        }
    }
}

// Quill editor styles

.ql-snow {
    & .ql-stroke {
        stroke: white;
        transition: .3s;
    }

    & .ql-fill {
        fill: white;
        transition: .3s;
    }

    & .ql-picker-label {
        color: white;

        &:hover {
            color: variables.$secondary;
        }
    }

    & .ql-picker-options {
        border: none !important;
        padding: 20px;
        background: variables.$black;
    }

    & .ql-picker-item {
        color: white;
        transition: .3s;

        &.ql-selected {
            color: variables.$secondary;
        }
    }

    & p {
        line-height: 40px;
        font-size: 18px;
        margin: 0;
    }

    &.ql-toolbar {
        &.disabled {
            display: none;
        }

        & .ql-picker-label:hover,
        & .ql-picker-item.ql-selected,
        & .ql-picker-item:hover {
            color: variables.$secondary;
        }

        & button:hover,
        & button:focus,
        & button.ql-active {
            & .ql-stroke {
                stroke: variables.$secondary;
            }
        
            & .ql-fill {
                fill: variables.$secondary;
            }
        }
    }

    & .ql-syntax {
        font-weight: bold;
        font-family: consolas;
        font-size: 18px;
    }
}