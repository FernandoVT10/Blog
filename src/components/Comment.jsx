import "../styles/components/comment.scss";

export default ({ comment }) => {
    return (
        <div className="comment">
            <p className="comment__name">
                { comment.name }
            </p>
            <p className="comment__content">
                { comment.comment }
            </p>
        </div>
    );
};