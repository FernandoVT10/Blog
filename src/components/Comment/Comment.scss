@use "../../styles/variables";

.comment {
    margin-top: 25px;
    width: 100%;
    padding: 20px;
    background: variables.$black;

    &__name {
        color: white;
        font-size: 18px;
        font-weight: bold;
    }

    &__content {
        color: white;
        margin-bottom: 0;
    }
}