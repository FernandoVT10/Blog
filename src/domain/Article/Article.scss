.article {
    background: var(--primary);
    padding: 0;
    margin-top: 30px;
    color: var(--white);
    
    &__cover {
        width: 100%;
        height: 400px;
        background-position: center !important;
        background-attachment: fixed !important;
        background-size: cover !important;
    }

    &__content-container {
        padding: 30px;
    }

    &__title {
        font-weight: 600;
        font-size: 2em;
    }

    &__categories {
        margin-top: 10px;
        font-weight: bold;
        font-size: 0.8em;
        color: var(--secondary);
    }

    &__category {
        display: inline-block;
        color: var(--secondary);
        transition: .3s;

        &:hover {
            text-decoration: none;
            color: var(--white);
        }
    }

    &__loader {
        display: flex;
        align-items: center;
        justify-content: center;
        background: var(--primary);
        color: var(--secondary);
        margin-top: 30px;
    }

    &__not-found {
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        color: var(--secondary);
        background: var(--primary);
        font-weight: bold;
        margin-top: 30px;
        font-size: 1.5em;
    }

    &__404 {
        font-size: 5em;
        color: var(--white);
    }
}