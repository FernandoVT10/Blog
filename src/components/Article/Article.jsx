import Cover from "./Cover";
import Title from "./Title";
import Content from "./Content";
import Categories from "./Categories";
import Description from "./Description";
import Api from "../../ApiController";

import { useState, useEffect } from "react";

import "./Article.scss";

export default (props) => {
    const [coverImage, setCoverImage] = useState();
    const [title, setTitle] = useState("");
    const [content, setContent] = useState("");
    const [description, setDescription] = useState("");
    const [categories, setCategories] = useState([]);

    const [article, setArticle] = useState(props.article);
    const [editArticle, setEditArticle] = useState(false);

    useEffect(() => {
        setTitle(article.title);
        setContent(article.content);
        setDescription(article.description);
        setCategories(article.categories.map(
            category => category.name
        ));
    }, []);

    const saveArticle = () => {
        if(!title.length) {
            return;
        }

        const formData = new FormData();

        formData.append("articleId", article._id);
        formData.append("cover", coverImage);
        formData.append("title", title);
        formData.append("content", content);
        formData.append("description", description);
        formData.append("categories", categories);

        Api.post("articles/updateArticle/", formData, true, true)
        .then(res => {
            if(res.status) {
                const newArticle = res.newArticle;

                setArticle(newArticle);
            }
            setEditArticle(false);
        });
    };

    const getFloatButtons = () => {
        if(editArticle) {
            return (
                <div className="article__buttons-container">
                    <button
                    className="article__float-button"
                    onClick={saveArticle}>
                        <i className="fas fa-folder-open" aria-hidden="true"></i>
                    </button>

                    <button
                    className="article__float-button article__float-button--cancel"
                    onClick={() => setEditArticle(false)}>
                        <i className="fas fa-times" aria-hidden="true"></i>
                    </button>
                </div>
            );
        }

        return (
            <div className="article__buttons-container">
                <button
                className="article__float-button"
                onClick={() => setEditArticle(true)}>
                    <i className="fas fa-pencil-alt" aria-hidden="true"></i>
                </button>
            </div>
        );
    };

    return (
        <div className="container-fluid">
            <div className="row">
                <div className="col-12 article">
                    <Cover
                    editArticle={editArticle}
                    cover={article.cover}
                    onChangeImage={setCoverImage}/>

                    <div className="article__content-container">
                        <Title
                        editArticle={editArticle}
                        title={article.title}
                        onChangeTitle={setTitle}/>
                        
                        <Content
                        editArticle={editArticle}
                        content={article.content}
                        onChangeContent={setContent}/>

                        <Description
                        editArticle={editArticle}
                        description={description}
                        onChangeDescription={setDescription}/>

                        <Categories
                        editArticle={editArticle}
                        categories={article.categories}
                        onChangeCategories={setCategories}/>
                    </div>
                </div>

                { getFloatButtons() }
            </div>
        </div>
    );
};