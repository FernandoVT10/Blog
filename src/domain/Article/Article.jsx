import Content from "./Content";

import ApiController from "../../services/ApiController";

import { useState, useEffect } from "react";
import { useRouter } from "next/router";
import Link from "next/link";

import "./Article.scss";

export default () => {
    const [article, setArticle] = useState({});
    const [loading, setLoading] = useState(true);

    const router = useRouter();

    useEffect(() => {
        const { articleId } = router.query;

        if(articleId && loading) {
            ApiController.get(`articles/${articleId}`)
            .then(res => {
                if(res.data) {
                    const { title } = res.data.article;
                    window.document.title = `${title} - Fernando Vaca Tamayo Blog`;

                    setArticle(res.data.article);
                }

                setLoading(false);
            });
        }
    }, [router.query]);

    const getCategories = () => {
        if(article.categories.length) {
            return (
                <div className="article__categories">
                    <i className="fas fa-tags" aria-hidden="true"></i>

                    <span className="ml-2">
                        {article.categories.map((category, index) => {
                            return (
                                <Link
                                href={`/articles?categories=${category.name}`}
                                key={index}>
                                    <a className="article__category mr-2">
                                        { category.name }
                                    </a>
                                </Link>
                            );
                        })}
                    </span>
                </div>
            );
        }
    }

    if(!article.title && !loading) {
        return (
            <div className="container-fluid">
                <div className="row">
                    <div className="col-12 body article__not-found">
                        <span className="article__404">404</span>
                        The article doesn't exist.
                    </div>
                </div>
            </div>
        );
    }

    if(loading) {
        return (
            <div className="container-fluid">
                <div className="row">
                    <div className="col-12 article__loader body">
                        <span className="spinner-border"></span>
                    </div>
                </div>
            </div>
        );
    }

    return (
        <div className="container-fluid">
            <div className="row">
                <div className="col-12 article">
                    <div
                    className="article__cover"
                    style={{ background: `url(/img/articles/${article.cover})` }}></div>

                    <div className="article__content-container">
                        <h1 className="article__title">
                            { article.title }
                        </h1>

                        <Content content={article.content}/>

                        { getCategories() }
                    </div>
                </div>
            </div>
        </div>
    );
};