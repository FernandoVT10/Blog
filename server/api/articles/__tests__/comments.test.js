import Comment from "../../../models/Comment";
import Article from "../../../models/Article";
import supertest from "supertest";
import app from "../../../app";

const request = supertest(app);

const COMMENT_MOCKS = [
    {
        name: "Test name",
        comment: "Test comment",
        createdAt: Date.now() + 2
    },
    {
        name: "Test name 2",
        comment: "Test comment 2",
        createdAt: Date.now() + 1
    },
    {
        name: "Test name 3",
        comment: "Test comment 3",
        createdAt: Date.now()
    }
];

const ARTICLE_MOCK = {
    title: "Test Article",
    description: "Test description",
    cover: "cover.jpg",
    content: "Test content",
    categories: [],
    createdAt: Date.now() + 2
};

setupTestDB("test_comments");

let article;

describe("comments api", () => {
    beforeEach(async () => {
        article = await Article.create(ARTICLE_MOCK);

        for(const { name, comment, createdAt } of COMMENT_MOCKS) {
            await Comment.create({
                articleId: article._id,
                name,
                comment,
                createdAt
            });
        }
    });

    afterEach(() => {
        article = null;
    });

    describe("Get Comments", () => {
        it("it should get the all comments", async () => {
            const res = await request.get(`/api/articles/${article._id}/comments`);
            const { comments } = res.body.data;

            comments.forEach(({ name, comment }, index) => {
                expect(name).toBe(COMMENT_MOCKS[index].name);
                expect(comment).toBe(COMMENT_MOCKS[index].comment);
            });
        });

        it("it should get the second comment", async () => {
            const res = await request
                .get(`/api/articles/${article._id}/comments`)
                .query({ offset: 1, limit: 1 });

            const { comments } = res.body.data;

            expect(comments[0].name).toBe("Test name 2");
            expect(comments[0].comment).toBe("Test comment 2");
        });

        it("it should get an empty array", async () => {
            const res = await request.get(`/api/articles/abcefabcefabcefabcefabce/comments`);
            const { comments } = res.body.data;

            expect(comments).toEqual([]);
        });
    });

    describe("Add Comment", () => {
        it("It should get a 404 error", async () => {
            const res = await request
                .post(`/api/articles/abcefabcefabcefabcefabce/comments`)
                .send({
                    name: "Test name",
                    comment: "Test comment"
                });

            expect(res.body).toEqual({
                errors: [
                    {
                        status: 404,
                        message: "The article doesn't exist"
                    }
                ]
            });
        });

        it("It should get a max character error in the name", async () => {
            const res = await request
                .post(`/api/articles/${article._id}/comments`)
                .send({
                    name: "Test name with more of 30 characters",
                    comment: "Test comment"
                });

            const { errors } = res.body;

            expect(errors).toMatchSnapshot();
        });

        it("It should get a max character error in the comment", async () => {
            const comment = randomText(501);

            const res = await request
                .post(`/api/articles/${article._id}/comments`)
                .send({
                    name: "Test name",
                    comment
                });

            const { errors } = res.body;

            expect(errors[0].message)
            .toBe(`comments validation failed: comment: Path \`comment\` (\`${comment}\`) is longer than the maximum allowed length (500).`);
        });

        it("It should create a comment", async () => {
            const comment = randomText(250);

            const res = await request
                .post(`/api/articles/${article._id}/comments`)
                .send({
                    name: "Test name #21",
                    comment
                });

            const { createdComment } = res.body.data;

            expect(createdComment.name).toBe("Test name #21");
            expect(createdComment.comment).toBe(comment);
        });
    });
});