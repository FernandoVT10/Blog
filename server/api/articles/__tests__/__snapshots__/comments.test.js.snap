// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comments api Add Comment It should get a max character error in the name 1`] = `
Array [
  Object {
    "_message": "comments validation failed",
    "errors": Object {
      "name": Object {
        "kind": "maxlength",
        "message": "Path \`name\` (\`Test name with more of 30 characters\`) is longer than the maximum allowed length (30).",
        "name": "ValidatorError",
        "path": "name",
        "properties": Object {
          "maxlength": 30,
          "message": "Path \`name\` (\`Test name with more of 30 characters\`) is longer than the maximum allowed length (30).",
          "path": "name",
          "type": "maxlength",
          "value": "Test name with more of 30 characters",
        },
        "value": "Test name with more of 30 characters",
      },
    },
    "message": "comments validation failed: name: Path \`name\` (\`Test name with more of 30 characters\`) is longer than the maximum allowed length (30).",
    "name": "ValidationError",
  },
]
`;
